src/sparseml/exporters/transforms/utils/helpers.py

```diff
@@ -161,3 +161,115 @@ def assert_node_type(node: NodeProto, op: Union[List[str], Set[str], str]) -> bo
         return node.op_type == op
     else:
         return node.op_type in op
+
+
+NodeOrInit = Union[NodeProto, TensorProto]
+
+
+class MatchResult:
+    def __init__(self, node: NodeOrInit) -> None:
+        self.node: Optional[NodeOrInit] = node
+        self.parents: List[Optional[NodeOrInit]] = []
+        self.children: List[Optional[NodeOrInit]] = []
+
+
+INITIALIZER_MATCH = "__Initializer__"
+
+
+def optional_node(tag: str) -> str:
+    return "Optional-" + tag
+
+
+def _is_optional_node(tag: str) -> Tuple[str, bool]:
+    if tag.startswith("Optional-"):
+        return tag.split("-")[1], True
+    else:
+        return tag, False
+
+
+def iter_structural_matches(
+    graph: ONNXGraph,
+    op_type: Optional[str] = None,
+    parent_ops: Optional[List[List[str]]] = None,
+    children_ops: Optional[List[List[str]]] = None,
+) -> Iterable[MatchResult]:
+    for node in graph._model.graph.node:
+        match = match_structure(
+            graph,
+            node,
+            op_type=op_type,
+            parent_ops=parent_ops,
+            children_ops=children_ops,
+        )
+        if match is not None:
+            yield match
+
+
+def match_structure(
```

Comment:
This is an awesome function, and my love of it is unjustifiably diminished by its complexity. Complexity is an easy thing to get rid of tho.

I know you probably now aimed for making the functionality work and will work on the cleanup now. Anyway, I'd like to propose:
- moving all the matching functions (incl. `iter_structural_matches` etc) to a separate script `.../exporters/transforms/utils/matching.py`
- breaking up `match_structure` into three subfunctions: `assert_node_validity` (L218-233), `assert_parent_ops` (L233-254) and `assert_children_ops (L254-275)
- adding comments

Reply:
Separated it out as you suggested and put it into matching.py